src/Host/Client/Impl/Host/RHost.cs

```diff
@@ -396,16 +396,16 @@ public sealed partial class RHost : IDisposable, IRExpressionEvaluator, IRBlobSe
                     await _transport.CloseAsync();
                     token.Set();
                 } catch (OperationCanceledException) {
+                    token.Set();
                 } catch (MessageTransportException) {
+                    token.Set();
                 } finally {
                     token.Reset();
```

Comment:
No. `BinaryAsyncLock` is kind of a mix of `ManualResetEvent`, `AutoResetEvent` and a boolean flag. If you call `Set`, all waiting threads are released and lock is set. It usually means that operation that has happened under lock is completed successfully and other threads don't need to repeat it. If you call `Reset`, only one waiting thread will be released and lock is not set. That means that old lock holder has failed to complete the operation, and the new lock holder need to make another attempt.
You can call `Set` or `Reset` multiple times, but only the first call matters. It is similar to `TaskCompletionSource.TrySetResult` and `TaskCompletionSource.TrySetCanceled` - you can't convert completed task into canceled and visa versa. To change `BinaryAsyncLock` state again, you either need to acquire another token or call `EnqueueReset`.